### DISCUSSION TITLE
Launch custom kitten directly, i.e. without key binding

### DISCUSSION THREAD

**Discussion by a3lem** [kovidgoyal/kitty#6693] [Q&A] [answered]
I'm just getting started writing my own kittens.

It seems I need to bind my new kitten to a keyboard shortcut in order to be able to run it. Is that correct? I'd be interested in being able to run my kittens just like like the built-in kittens, that is, directly from the command line. I was kind of surprised I can't get this to work.

First attempt:

```
kitten my_kitty.py
```

Second:

```
kitten my_kitty
```

Read the error message: 

```
my_kitty.py is not a known subcommand for kitten
```

Alright, that explains it, I suppose.

And so I searched elsewhere. Seeing as there's a `+runpy` subcommand I checked if `+run-kitten` was in there too. Nope. I tried launching the kitten with `+launch`. That didn't do anything.

These dead ends lead me to assume that it's keymaps or nothing. Am I maybe missing something? Thanks!

**Comment by kovidgoyal** [comment, marked as answer]
What would be the point of running a custom kitten from the command line? At
that point its just a regular program, write it as a regular program and
run it as a regular program. The various builtin kittens that operate as
regular programs just use the kitten infrastructure as a unified
launching point, that's all. A custom kitten comes in handy only when it
has the handle_result() function. And this function must be run in the
kitty process and so must be mapped. If you want to run it by typing
instead, you can start a kitty remote control shell (ctrl+shift+esc)
and type

kitten mykitten.py

and press enter

If you wish to write a python script making use of kitty infrastructure
to run as a regular program, you can use

kitty +launch myscript.py

On Mon, Oct 09, 2023 at 01:11:27PM -0700, adriaan wrote:
> I'm just getting started writing my own kittens.
> 
> It seems I need to bind them to a keyboard shortcut in order to be able to run them. Is that correct? I'd like to be able to run my kittens just like like the built-in kittens, that is, directly from the command line. I was kind of surprised this doesn't work:
> 
> ```
> kitten my_kitty.py
> ```
> 
> I also tried looking for something like this:
> 
> ```
> kitty +run-kitten my_kitty.py
> ```
> 
> Am I maybe missing something? Thanks!
> 
> 
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6693
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Reply by a3lem** [reply]
Thanks for clarifying. That actually makes a lot of sense.

I made a silly mistake when testing the `kitty +launch` approach: I forgot to actually *call* one of the functions in the script. That's why it seemed to 'do nothing'. (My bad!) But now that I'm running the script correctly, I can confirm that it does have access to the kitty infrastructure: I'm able to import from the `kitty` package.

Great! Resolved.

---

If I may ask a follow-up question, is there a way for me to point my IDE/editor at Kitty's Python environment? I'm on macOS. 

The command...

```
kitty +runpy "import sys; print(sys.exec_prefix)"
``` 

... returns `/Applications/kitty.app/Contents/Resources/Python/lib/python3.9`. There doesn't seem to be an actual file at that path though.

Maybe I need to add something to my PYTHONPATH instead?

I'm curious how you or others do this? Or does everyone just put with the error messages about package `kitty` not being resolvable?

Thanks!

**Reply by kovidgoyal** [reply]
Install the kitty source code and point your IDE at it.
